go/chat/localizer.go

```diff
@@ -0,0 +1,952 @@
+			select {
+			case <-ctx.Done():
+				s.Debug(ctx, "localizeConversations: context is done, bailing (producer)")
+				return ctx.Err()
+			default:
+			}
+			convCh <- job{conv: conv, index: i}
+		}
+		return nil
+	})
+	nthreads := s.G().Env.GetChatInboxSourceLocalizeThreads()
+	for i := 0; i < nthreads; i++ {
+		index := i
+		eg.Go(func() error {
+			for conv := range convCh {
+				s.gateCheck(ctx, localizeJob.gateCh, index)
+				s.Debug(ctx, "localizeConversations: localizing: %d convID: %s", index, conv.conv.GetConvID())
+				convLocal := s.localizeConversation(ctx, uid, conv.conv)
+				select {
+				case <-ctx.Done():
+					s.Debug(ctx, "localizeConversations: context is done, bailing (consumer): %d", index)
+					return ctx.Err()
+				default:
+				}
+				retCh <- conv.conv.GetConvID()
+				if convLocal.Error != nil {
+					s.Debug(ctx, "localizeConversations: error localizing: convID: %s err: %s",
+						conv.conv.GetConvID(), convLocal.Error.Message)
+				}
+				localizeJob.retCh <- types.AsyncInboxResult{
+					ConvLocal: convLocal,
+					Conv:      conv.conv,
+				}
+			}
+			return nil
+		})
+	}
+	go func() {
+		eg.Wait()
+		close(retCh)
+	}()
+	for convID := range retCh {
+		localizeJob.complete(convID)
+	}
+	return eg.Wait()
+}
+
+func (s *localizerPipeline) isErrPermanent(err error) bool {
+	if uberr, ok := err.(UnboxingError); ok {
+		return uberr.IsPermanent()
+	}
+	return false
+}
+
+func getUnverifiedTlfNameForErrors(conversationRemote chat1.Conversation) string {
+	var tlfName string
+	var latestMsgID chat1.MessageID
+	for _, msg := range conversationRemote.MaxMsgSummaries {
+		if msg.GetMessageID() > latestMsgID {
+			latestMsgID = msg.GetMessageID()
+			tlfName = msg.TLFNameExpanded(conversationRemote.Metadata.FinalizeInfo)
+		}
+	}
+	return tlfName
+}
+
+func (s *localizerPipeline) getMessagesOffline(ctx context.Context, convID chat1.ConversationID,
+	uid gregor1.UID, msgs []chat1.MessageSummary, finalizeInfo *chat1.ConversationFinalizeInfo) ([]chat1.MessageUnboxed, chat1.ConversationErrorType, error) {
+
+	st := storage.New(s.G(), s.G().ConvSource)
+	res, err := st.FetchMessages(ctx, convID, uid, utils.PluckMessageIDs(msgs))
+	if err != nil {
+		// Just say we didn't find it in this case
+		return nil, chat1.ConversationErrorType_TRANSIENT, err
+	}
+
+	// Make sure we got legit msgs
+	var foundMsgs []chat1.MessageUnboxed
+	for _, msg := range res {
+		if msg != nil {
+			foundMsgs = append(foundMsgs, *msg)
+		}
+	}
+
+	if len(foundMsgs) == 0 {
+		return nil, chat1.ConversationErrorType_TRANSIENT, errors.New("missing messages locally")
+	}
+
+	return foundMsgs, chat1.ConversationErrorType_NONE, nil
+}
+
+func (s *localizerPipeline) getMinWriterRoleInfoLocal(ctx context.Context, info *chat1.ConversationMinWriterRoleInfo) *chat1.ConversationMinWriterRoleInfoLocal {
+	if info == nil {
+		return nil
+	}
+	username := ""
+	name, err := s.G().GetUPAKLoader().LookupUsername(ctx, keybase1.UID(info.Uid.String()))
```

Comment:
realize this is just a mv but should we use uid mapper here instead for performance?